framework/Models/HybridModels/HybridModelBase.py

```diff
@@ -0,0 +1,229 @@
+  """
+
+  @classmethod
+  def getInputSpecification(cls):
+    """
+      Method to get a reference to a class that specifies the input data for the class cls.
+      @ In, cls, the class for which we are retrieving the specification
+      @ Out, inputSpecification, InputData.ParameterInput, class to use for specifying input of cls.
+    """
+    inputSpecification = super(HybridModelBase, cls).getInputSpecification()
+    modelInput = InputData.parameterInputFactory("Model", contentType=InputTypes.StringType)
+    modelInput.addParam("class", InputTypes.StringType)
+    modelInput.addParam("type", InputTypes.StringType)
+    inputSpecification.addSub(modelInput)
+    return inputSpecification
+
+  @classmethod
+  def specializeValidateDict(cls):
+    """
+      This method describes the types of input accepted with a certain role by the model class specialization
+      @ In, None
+      @ Out, None
+    """
+    pass
+
+  def __init__(self,runInfoDict):
+    """
+      Constructor
+      @ In, runInfoDict, dict, the dictionary containing the runInfo (read in the XML input file)
+      @ Out, None
+    """
+    Dummy.__init__(self,runInfoDict)
+    self.modelInstances        = {}                  # dictionary {modelName: modelInstance}: instances of given model
+    self.sleepTime             = 0.005               # waiting time before checking if a run is finished.
+    self.printTag              = 'HybridModelBase MODEL' # print tag
+    self.createWorkingDir      = False               # If the type of model is 'Code', this will set to true
+    # assembler objects to be requested
+    self.addAssemblerObject('Model','n',True)
+
+  def localInputAndChecks(self,xmlNode):
+    """
+      Function to read the portion of the xml input that belongs to this specialized class
+      and initialize some stuff based on the inputs got
+      @ In, xmlNode, xml.etree.ElementTree.Element, Xml element node
+      @ Out, None
+    """
+    Dummy.localInputAndChecks(self, xmlNode)
+    paramInput = self.getInputSpecification()()
+    paramInput.parseNode(xmlNode)
+    for modelNode in paramInput.findAll('Model'):
+      self.modelInstances.update({modelNode.value: None})
+      if not self.createWorkingDir and modelNode.parameterValues['type'] == 'Code':
+        self.createWorkingDir = True
+
+  def initialize(self,runInfo,inputs,initDict=None):
+    """
+      Method to initialize this model class
+      @ In, runInfo, dict, is the run info from the jobHandler
+      @ In, inputs, list, is a list containing whatever is passed with an input role in the step
+      @ In, initDict, dict, optional, dictionary of all objects available in the step is using this model
+      @ Out, None
+    """
+    for model in self.modelInstances:
+      if isinstance(model, Models.Model):
+        self.raiseAnError(IOError, "Model {} has already been initialized, and it can not be initialized again!".format(model.name))
```

Comment:
This is interesting. I was under the impression each entity gets initialized once per Step that it's used in; does this mean a model used in a Step and then later as part of a Hybrid will error out?

Reply:
This means if HybridModel is already initialized , it will error out, not specific model. Basically, I'm limiting the use of HybridModel in different steps. When it is moved forward to other models, i.e. LogicalModel, it will be more clear to me how to handle this. This is just a restructure with the same capabilities.

Reply:
Is this valid for any model or just for the ROM? I think we added this in the Hybrid for the ROM to be sure to not perform a "destructive" process in case of already trained ROM. In here, is it going to error out for whatever model?

Reply:
@alfoa You are right, it is used for the ROM. As I said, I prefer to change it in the next PR.